File: sdk/python/src/anyvali/sensitive.py
"""Explicit encrypted-storage helpers for ``sensitive`` schema nodes."""

from typing import Any, Callable, TypeVar

from .schemas.base import BaseSchema, ValidationContext
from .types import ParseResult, ValidationError

T = TypeVar("T")
SensitiveTransform = Callable[[list[str | int], Any], Any]


def _transform_sensitive(
    schema: BaseSchema,
    data: Any,
    mode: str,
    transform: SensitiveTransform,
) -> Any:
    result = schema._safe_parse_with_context(
        data,
        ValidationContext(
            sensitive_mode=mode,  # type: ignore[arg-type]
            sensitive_transform=transform,
            sensitive_cache={},
        ),
    )
    if not result.success:
        raise ValidationError(result.issues)
    return result.data


def safe_parse_encrypted(schema: BaseSchema, data: Any) -> ParseResult[Any]:
    """Validate encrypted storage data without decrypting sensitive values."""
    return schema._safe_parse_with_context(
        data, ValidationContext(sensitive_mode="encrypted")
    )


def encrypt(schema: BaseSchema, data: Any, transform: SensitiveTransform) -> Any:
    """Parse plaintext, transform sensitive values, then validate storage data."""
    encrypted = _transform_sensitive(schema, schema.parse(data), "encrypt", transform)
    result = safe_parse_encrypted(schema, encrypted)
    if not result.success:
        raise ValidationError(result.issues)
    return result.data


def decrypt(schema: BaseSchema[T], data: Any, transform: SensitiveTransform) -> T:
    """Validate storage data, transform sensitive values, then parse plaintext."""
    encrypted = safe_parse_encrypted(schema, data)
    if not encrypted.success:
        raise ValidationError(encrypted.issues)
    return schema.parse(_transform_sensitive(schema, encrypted.data, "decrypt", transform))

File: sdk/js/src/sensitive.ts
import { ValidationError } from "./errors.js";
import type { ParseResult, SensitiveTransform } from "./types.js";
import { BaseSchema } from "./schemas/base.js";

function transformSensitive(
  schema: BaseSchema,
  data: unknown,
  mode: "encrypt" | "decrypt",
  transform: SensitiveTransform,
): unknown {
  const result = schema._safeParseWithContext(data, {
    path: [],
    issues: [],
    sensitiveMode: mode,
    sensitiveTransform: transform,
    sensitiveCache: new Map(),
  });
  if (!result.success) throw new ValidationError(result.issues);
  return result.data;
}

/** Validate encrypted storage data without decrypting sensitive values. */
export function safeParseEncrypted(
  schema: BaseSchema,
  data: unknown,
): ParseResult<unknown> {
  return schema._safeParseWithContext(data, {
    path: [],
    issues: [],
    sensitiveMode: "encrypted",
  });
}

/** Parse plaintext, transform sensitive values, then validate the encrypted result. */
export function encrypt(
  schema: BaseSchema,
  data: unknown,
  transform: SensitiveTransform,
): unknown {
  const encrypted = transformSensitive(schema, schema.parse(data), "encrypt", transform);
  const result = safeParseEncrypted(schema, encrypted);
  if (!result.success) throw new ValidationError(result.issues);
  return result.data;
}

/** Validate encrypted data, transform sensitive values, then parse the plaintext result. */
export function decrypt<T>(
  schema: BaseSchema<unknown, T>,
  data: unknown,
  transform: SensitiveTransform,
): T {
  const encrypted = safeParseEncrypted(schema, data);
  if (!encrypted.success) throw new ValidationError(encrypted.issues);
  return schema.parse(transformSensitive(schema, encrypted.data, "decrypt", transform));
}
